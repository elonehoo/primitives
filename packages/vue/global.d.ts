import type { AxeMatchers } from 'vitest-axe/matchers'
/// <reference types="vite/client" />
import 'vitest'

declare module 'vitest' {
  export interface AsymmetricMatchersContaining extends AxeMatchers {}
  export interface Assertion extends AxeMatchers {}
}
