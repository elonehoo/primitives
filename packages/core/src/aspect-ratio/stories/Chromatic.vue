<script setup lang="ts">
import { AspectRatio } from '../index.ts'
import Img from './Img.vue'
import './styles.css'
</script>

<template>
  <div>
    <h1>Default ratio</h1>
    <div :style="{ width: '300px' }">
      <AspectRatio class="aspectRation__root">
        <p>Default ratio (1/1)</p>
      </AspectRatio>
    </div>

    <h1>Custom ratios</h1>
    <div :style="{ display: 'flex', gap: '20px' }">
      <div :style="{ width: '200px' }">
        <AspectRatio :ratio="1 / 2">
          <Img />
        </AspectRatio>
      </div>
      <div :style="{ width: '200px' }">
        <AspectRatio>
          <Img />
        </AspectRatio>
      </div>
      <div :style="{ width: '200px' }">
        <AspectRatio :ratio="16 / 9">
          <Img />
        </AspectRatio>
      </div>
      <div :style="{ width: '200px' }">
        <AspectRatio :ratio="2 / 1">
          <Img />
        </AspectRatio>
      </div>
    </div>
  </div>
</template>
