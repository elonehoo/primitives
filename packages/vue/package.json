{
  "name": "vue-packages",
  "type": "module",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "dev": "vite",
    "build": "vite build",
    "preview": "vite preview"
  },
  "dependencies": {
    "@rollup/plugin-alias": "^5.1.1"
  },
  "devDependencies": {
    "@vitejs/plugin-vue": "^5.1.4",
    "@vueuse/core": "^10.11.1",
    "filter-anything": "^3.0.7",
    "typescript": "^5.6.3",
    "unplugin-vue-components": "^0.26.0",
    "vite-plugin-pages": "^0.32.3",
    "vue-router": "^4.4.5",
    "vue-tsc": "^2.1.6"
  }
}
