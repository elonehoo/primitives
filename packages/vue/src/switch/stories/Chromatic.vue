<script setup lang="ts">
import { OkuSwitch, OkuSwitchThumb } from '@oku-ui/switch'
</script>

<template>
  <h1>Uncontrolled</h1>
  <h2>Off</h2>
  <OkuSwitch class="switch">
    <OkuSwitchThumb class="switch-thumb" />
  </OkuSwitch>

  <h2>On</h2>
  <OkuSwitch class="switch" default-checked>
    <OkuSwitchThumb class="switch-thumb" />
  </OkuSwitch>

  <h1>Controlled</h1>
  <h2>Off</h2>
  <OkuSwitch class="switch" :checked="false">
    <OkuSwitchThumb class="switch-thumb" />
  </OkuSwitch>

  <h2>On</h2>
  <OkuSwitch class="switch" checked>
    <OkuSwitchThumb class="switch-thumb" />
  </OkuSwitch>

  <h1>Disabled</h1>
  <OkuSwitch class="switch" disabled>
    <OkuSwitchThumb class="switch-thumb" />
  </OkuSwitch>

  <h1>State attributes</h1>
  <h2>Unchecked</h2>
  <OkuSwitch class="switch-attr-styles">
    <OkuSwitchThumb class="switch-attr-styles" />
  </OkuSwitch>

  <h2>Checked</h2>
  <OkuSwitch class="switch-attr-styles" default-checked>
    <OkuSwitchThumb class="switch-attr-styles" />
  </OkuSwitch>

  <h2>Disabled</h2>
  <OkuSwitch class="switch-attr-styles" default-checked disabled>
    <OkuSwitchThumb class="switch-attr-styles" />
  </OkuSwitch>
</template>
