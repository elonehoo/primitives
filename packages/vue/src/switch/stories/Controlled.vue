<script setup lang="ts">
import { ref } from 'vue'
import { OkuSwitch, OkuSwitchThumb } from '@oku-ui/switch'
import Label from './Label.vue'

const checked = ref<boolean>(true)
</script>

<template>
  <p>This switch is placed adjacent to its label. The state is controlled.</p>
  <Label for="randBox">This is the label</Label>{{ ' ' }}
  <OkuSwitch
    id="randBox"
    class="switch"
    :checked="checked"
    @checked-change="checked = $event"
  >
    <OkuSwitchThumb class="switch-thumb" />
  </OkuSwitch>
</template>
