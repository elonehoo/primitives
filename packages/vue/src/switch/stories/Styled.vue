<script setup lang="ts">
import { OkuSwitch, OkuSwitchThumb } from '@oku-ui/switch'
import Label from './Label.vue'
</script>

<template>
  <p>This switch is nested inside a label. The state is uncontrolled.</p>
  <Label>
    This is the label{{ ' ' }}
    <OkuSwitch class="switch">
      <OkuSwitchThumb class="switch-thumb" />
    </OkuSwitch>
  </Label>
</template>
