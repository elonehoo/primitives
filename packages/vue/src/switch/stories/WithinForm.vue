<script setup lang="ts">
import { ref } from 'vue'
import { OkuSwitch, OkuSwitchThumb } from '@oku-ui/switch'

const data = ref({ optional: false, required: false, stopprop: false })
const checked = ref<boolean>(false)
</script>

<template>
  <form
    @submit="(event) => event.preventDefault()"
    @change="(event) => {
      // TODO: this does not work since the `bubble-input` does not bubble checked change to parents, so we used v-model instated.
      const input = event.target as HTMLInputElement
      data = { ...data, [input.name]: input.checked }
    }"
  >
    <fieldset>
      <legend>optional checked: {{ String(data.optional) }}</legend>
      <label>
        <OkuSwitch
          class="switch"
          name="optional"
          :checked="checked"
          @checked-change="checked = $event"
        >
          <OkuSwitchThumb class="switch-thumb" />
        </OkuSwitch>{{ ' ' }}
        with label
      </label>
    </fieldset>

    <br>
    <br>

    <fieldset>
      <legend>required checked: {{ String(data.required) }}</legend>
      <OkuSwitch class="switch" name="required" required>
        <OkuSwitchThumb class="switch-thumb" />
      </OkuSwitch>
    </fieldset>

    <br>
    <br>

    <fieldset>
      <legend>stop propagation checked: {{ String(data.stopprop) }}</legend>
      <OkuSwitch
        class="switch"
        name="stopprop"
        @click="(event) => event.stopPropagation()"
      >
        <OkuSwitchThumb class="switch-thumb" />
      </OkuSwitch>
    </fieldset>

    <br>
    <br>

    <button>Submit</button>
  </form>
</template>
