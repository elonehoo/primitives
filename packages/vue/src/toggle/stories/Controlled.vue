<script setup lang="ts">
import { OkuToggle } from '@oku-ui/toggle'
import { ref } from 'vue'

const pressed = ref(true)
</script>

<template>
  <OkuToggle class="toggle" :pressed="pressed" @pressed-change="pressed = $event">
    {{ pressed ? 'On' : 'Off' }}
  </OkuToggle>
</template>
