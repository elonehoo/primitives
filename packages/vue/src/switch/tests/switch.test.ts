import { defineComponent } from 'vue'
import { enableAutoUnmount, mount, shallowMount } from '@vue/test-utils'
import type { DOMWrapper, VueWrapper } from '@vue/test-utils'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { axe } from 'vitest-axe'

import { OkuSwitch, OkuSwitchThumb } from '../'

import Styled from '..//stories/Styled.vue'
import Controlled from '../stories/Controlled.vue'
import WithinForm from '../stories/WithinForm.vue'
import Chromatic from '../stories/Chromatic.vue'

enableAutoUnmount(afterEach)

const SWITCH_ROLE = 'switch'
const THUMB_TEST_ID = 'switch-indicator'

const onCheckedChange = vi.fn()

const SwitchTest = defineComponent({
  components: {
    OkuSwitch,
    OkuSwitchThumb,
  },
  props: {
    defaultChecked: Boolean,
    checked: Boolean,
    disabled: Boolean,
  },
  setup() {
    return {
      THUMB_TEST_ID,
      onCheckedChange,
    }
  },
  template: `
    <OkuSwitch aria-label="basic switch" v-bind="$attrs" @checked-change="onCheckedChange">
      <OkuSwitchThumb :data-testid="THUMB_TEST_ID" />
    </OkuSwitch>
  `,
})

describe('okuSwitch', () => {
  let wrapper: VueWrapper
  let switcher: DOMWrapper<Element>
  let thumb: DOMWrapper<Element>

  beforeEach(() => {
    wrapper = mount(SwitchTest, {
      attachTo: document.body,
    })

    switcher = wrapper.find(`[role="${SWITCH_ROLE}"]`)
    thumb = wrapper.find(`[data-testid="${THUMB_TEST_ID}"]`)
  })

  it('should render OkuSwitch correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render OkuSwitchThumb correctly', () => {
    const spy = vi.spyOn(globalThis.console, 'warn').mockImplementation(() => { })
    const wrapper = () => mount(OkuSwitchThumb)

    expect(() => wrapper()).toThrowErrorMatchingSnapshot()

    expect(spy).toHaveBeenCalled()

    expect(spy.mock.calls[0][0]).toContain('[Vue warn]: injection "Symbol(OkuSwitch)" not found.')
  })

  /**
   * @vitest-environment jsdom
   */
  it('should have no accessibility violations', async () => {
    expect(await axe(wrapper.element)).toHaveNoViolations()
  })

  describe('when clicking the switch', () => {
    beforeEach(async () => {
      switcher.trigger('click')
    })

    it('should render OkuSwitch correctly with the correct attributes', () => {
      expect(switcher.exists()).toBe(true)
      expect(switcher.element.tagName).toBe('BUTTON')
      expect(switcher.attributes('type')).toBe('button')
      expect(switcher.attributes('role')).toBe('switch')
      expect(switcher.attributes('aria-checked')).toBe('true')
      expect(switcher.attributes('data-state')).toBe('checked')
      expect(switcher.attributes('value')).toBe('on')
    })

    it('should render OkuSwitchThumb correctly with the correct attributes', () => {
      expect(thumb.exists()).toBe(true)
      expect(thumb.element.tagName).toBe('SPAN')
      expect(thumb.attributes('data-state')).toBe('checked')
    })

    it('should call `onCheckedChange` event', () => {
      expect(onCheckedChange).toHaveBeenCalled()
    })

    describe('and clicking the switch again', () => {
      beforeEach(async () => {
        switcher.trigger('click')
      })

      it('should render OkuSwitch correctly with the correct attributes', () => {
        expect(switcher.exists()).toBe(true)
        expect(switcher.element.tagName).toBe('BUTTON')
        expect(switcher.attributes('type')).toBe('button')
        expect(switcher.attributes('role')).toBe('switch')
        expect(switcher.attributes('aria-checked')).toBe('false')
        expect(switcher.attributes('data-state')).toBe('unchecked')
        expect(switcher.attributes('value')).toBe('on')
      })

      it('should render OkuSwitchThumb correctly with the correct attributes', () => {
        expect(thumb.exists()).toBe(true)
        expect(thumb.element.tagName).toBe('SPAN')
        expect(thumb.attributes('data-state')).toBe('unchecked')
      })
    })
  })

  describe('given a disabled Switch', () => {
    let wrapper: VueWrapper

    beforeEach(() => {
      wrapper = mount(SwitchTest, {
        props: {
          disabled: false,
        },
        attachTo: document.body,
      })
    })

    /**
     * @vitest-environment jsdom
     */
    it('should have no accessibility violations', async () => {
      expect(await axe(wrapper.element)).toHaveNoViolations()
    })
  })

  describe('given an uncontrolled `checked` Switch', () => {
    let wrapper: VueWrapper
    let switcher: DOMWrapper<Element>
    let thumb: DOMWrapper<Element>

    beforeEach(() => {
      wrapper = mount(SwitchTest, {
        props: {
          defaultChecked: true,
        },
        attachTo: document.body,
      })

      switcher = wrapper.find(`[role="${SWITCH_ROLE}"]`)
      thumb = wrapper.find(`[data-testid="${THUMB_TEST_ID}"]`)
    })

    /**
     * @vitest-environment jsdom
     */
    it('should have no accessibility violations', async () => {
      expect(await axe(wrapper.element)).toHaveNoViolations()
    })

    describe('when clicking the switch', () => {
      beforeEach(async () => {
        switcher.trigger('click')
      })

      it('should render OkuSwitch correctly with the correct attributes', () => {
        expect(switcher.exists()).toBe(true)
        expect(switcher.element.tagName).toBe('BUTTON')
        expect(switcher.attributes('type')).toBe('button')
        expect(switcher.attributes('role')).toBe('switch')
        expect(switcher.attributes('aria-checked')).toBe('true')
        expect(switcher.attributes('data-state')).toBe('checked')
        expect(switcher.attributes('value')).toBe('on')
      })

      it('should render OkuSwitchThumb correctly with the correct attributes', () => {
        expect(thumb.exists()).toBe(true)
        expect(thumb.element.tagName).toBe('SPAN')
        expect(thumb.attributes('data-state')).toBe('checked')
      })

      it('should call `onCheckedChange` event', () => {
        expect(onCheckedChange).toHaveBeenCalled()
      })
    })

    describe('given a controlled `checked` Switch', () => {
      let wrapper: VueWrapper
      let switcher: DOMWrapper<Element>

      beforeEach(() => {
        wrapper = mount(SwitchTest, {
          props: {
            checked: true,
          },
          attachTo: document.body,
        })

        switcher = wrapper.find(`[role="${SWITCH_ROLE}"]`)
      })

      describe('when clicking the switch', () => {
        beforeEach(() => {
          switcher.trigger('click')
        })

        it('should call `onCheckedChange` event', () => {
          expect(onCheckedChange).toHaveBeenCalled()
        })
      })
    })
  })
})

describe('switch v-model tests', () => {
  it('checked', async () => {
    const wrapper = mount(OkuSwitch, {
      props: {
        'onUpdate:checked': (e: boolean) => wrapper.setProps({ checked: e }),
      },
    })

    await wrapper.find('button').trigger('click')
    expect(wrapper.props('checked')).toBe(true)

    await wrapper.find('button').trigger('click')
    expect(wrapper.props('checked')).toBe(false)
  })

  it('checked init true should be updated', async () => {
    const wrapper = mount(OkuSwitch, {

      props: {
        'checked': true,
        'onUpdate:checked': (e: boolean) => wrapper.setProps({ checked: e }),
      },
    })

    await wrapper.find('button').trigger('click')
    expect(wrapper.props('checked')).toBe(false)

    await wrapper.find('button').trigger('click')
    expect(wrapper.props('checked')).toBe(true)
  })
})

describe('okuSwitch Stories', () => {
  describe('styled', () => {
    let wrapper: VueWrapper<InstanceType<typeof Styled>>

    beforeEach(async () => {
      wrapper = shallowMount(Styled, {
        attachTo: document.body,
      })
    })

    /**
     * @vitest-environment jsdom
     */
    it('should have no accessibility violations', async () => {
      expect(await axe(wrapper.element)).toHaveNoViolations()
    })

    it('should render correctly', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('controlled', () => {
    let wrapper: VueWrapper<InstanceType<typeof Controlled>>

    beforeEach(async () => {
      wrapper = shallowMount(Controlled, {
        attachTo: document.body,
      })
    })

    /**
     * @vitest-environment jsdom
     */
    it('should have no accessibility violations', async () => {
      expect(await axe(wrapper.element)).toHaveNoViolations()
    })

    it('should render correctly', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('withinForm', () => {
    let wrapper: VueWrapper<InstanceType<typeof WithinForm>>

    beforeEach(async () => {
      wrapper = shallowMount(WithinForm, {
        attachTo: document.body,
      })
    })

    /**
     * @vitest-environment jsdom
     */
    it('should have no accessibility violations', async () => {
      expect(await axe(wrapper.element)).toHaveNoViolations()
    })

    it('should render correctly', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('chromatic', () => {
    let wrapper: VueWrapper<InstanceType<typeof Chromatic>>

    beforeEach(async () => {
      wrapper = shallowMount(Chromatic, {
        attachTo: document.body,
      })
    })

    /**
     * @vitest-environment jsdom
     */
    it('should have no accessibility violations', async () => {
      expect(await axe(wrapper.element)).toHaveNoViolations()
    })

    it('should render correctly', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
