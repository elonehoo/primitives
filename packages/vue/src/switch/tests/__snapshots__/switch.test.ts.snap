// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`okuSwitch > should render OkuSwitch correctly 1`] = `
"<button aria-label="basic switch" type="button" role="switch" aria-checked="false" aria-required="false" data-state="unchecked" value="on"><span data-testid="switch-indicator" data-state="unchecked"></span></button>
<!--v-if-->"
`;

exports[`okuSwitch > should render OkuSwitchThumb correctly 1`] = `[TypeError: Cannot read properties of undefined (reading 'is')]`;

exports[`okuSwitch Stories > chromatic > should render correctly 1`] = `
"<h1>Uncontrolled</h1>
<h2>Off</h2>
<oku-switch-stub required="false" disabled="false" is="button" aschild="false" class="switch" value="on"></oku-switch-stub>
<h2>On</h2>
<oku-switch-stub defaultchecked="true" required="false" disabled="false" is="button" aschild="false" class="switch" value="on"></oku-switch-stub>
<h1>Controlled</h1>
<h2>Off</h2>
<oku-switch-stub checked="false" required="false" disabled="false" is="button" aschild="false" class="switch" value="on"></oku-switch-stub>
<h2>On</h2>
<oku-switch-stub checked="true" required="false" disabled="false" is="button" aschild="false" class="switch" value="on"></oku-switch-stub>
<h1>Disabled</h1>
<oku-switch-stub required="false" disabled="true" is="button" aschild="false" class="switch" value="on"></oku-switch-stub>
<h1>State attributes</h1>
<h2>Unchecked</h2>
<oku-switch-stub required="false" disabled="false" is="button" aschild="false" class="switch-attr-styles" value="on"></oku-switch-stub>
<h2>Checked</h2>
<oku-switch-stub defaultchecked="true" required="false" disabled="false" is="button" aschild="false" class="switch-attr-styles" value="on"></oku-switch-stub>
<h2>Disabled</h2>
<oku-switch-stub defaultchecked="true" required="false" disabled="true" is="button" aschild="false" class="switch-attr-styles" value="on"></oku-switch-stub>"
`;

exports[`okuSwitch Stories > controlled > should render correctly 1`] = `
"<p>This switch is placed adjacent to its label. The state is controlled.</p>
<label-stub for="randBox"></label-stub>

<oku-switch-stub checked="true" required="false" disabled="false" is="button" aschild="false" id="randBox" class="switch" value="on"></oku-switch-stub>"
`;

exports[`okuSwitch Stories > styled > should render correctly 1`] = `
"<p>This switch is nested inside a label. The state is uncontrolled.</p>
<label-stub></label-stub>"
`;

exports[`okuSwitch Stories > withinForm > should render correctly 1`] = `
"<form>
  <fieldset>
    <legend>optional checked: false</legend><label>
      <oku-switch-stub name="optional" checked="false" required="false" disabled="false" is="button" aschild="false" class="switch" value="on"></oku-switch-stub> with label
    </label>
  </fieldset><br><br>
  <fieldset>
    <legend>required checked: false</legend>
    <oku-switch-stub name="required" required="true" disabled="false" is="button" aschild="false" class="switch" value="on"></oku-switch-stub>
  </fieldset><br><br>
  <fieldset>
    <legend>stop propagation checked: false</legend>
    <oku-switch-stub name="stopprop" required="false" disabled="false" is="button" aschild="false" class="switch" value="on"></oku-switch-stub>
  </fieldset><br><br><button>Submit</button>
</form>"
`;
